feat: move Message and Registration handlers to own validated class

jrconlin opened the pull request:
Continuing the effort to break up endpoint.py, move Message and
Registration handlers to their own marshmallow validated classes. I have
tried not to disturb tests unless absolutely required (e.g. @patch for
some reason stopped modifying `uuid.uuid4()`)

Closes #491

pjenvey commented on autopush/message/validation.py:
```diff
@@ -0,0 +1,43 @@
+                        channel_id=notif.channel_id,
+                        topic=notif.topic,
+                        message_id=message_id)
+        except (InvalidToken, InvalidTokenException):
```
The try block only needs to guard the from_message_id call, I think

pjenvey commented on autopush/truss/validation.py:
```diff
@@ -0,0 +1,116 @@
+            try:
+                uuid.UUID(uaid)
+            except (ValueError, TypeError):
+                raise InvalidRequest("", status_code=401, errno=109)
```
"Invalid uaid" and chid below, I suppose

pjenvey commented on autopush/truss/handler.py:
```diff
@@ -0,0 +1,234 @@
+
+        Endpoint generation and optionally router type/data registration.
+
+        :param truss_data: Dictionary of validated values.
```
the validator actually *doesn't* pass the validated dict as this argument, it only sets it as self.valid_input and passes along the args/kwargs from Cyclone. So RegistrationHandler is actually broken right now (Cyclone should still be calling it with 4 arguments)

Which reminds me, I would actually *like* the validator to work like that instead of setting self.valid_input. feels cleaner to me and AFAICT we're always passing all of the cyclone args through the validation step anyway.. passing them along unvalidated is confusing.

but let's just fix the args here for now and think about the validator change later

jrconlin replied:
Agreed. I should just mark the args as "ignored" since we don't use them anyway.

bbangert commented on autopush/base.py:
```diff
@@ -1,8 +1,122 @@
+    """Parsed notification from the request"""
+    version = attrib()
+    data = attrib()
+    channel_id = attrib()
```
If we're moving this around, maybe it should join WebPushNotification in utils? Or its own module?

bbangert commented on autopush/message/handler.py:
```diff
@@ -0,0 +1,42 @@
+from autopush.utils import WebPushNotification
+
+
+class MessageHandler(BaseWebHandler):
```
Why isn't this going to live in autopush/web? I'm trying to understand the architecture consistency being chosen. The intent with autopush/web was all web handlers would live there, like the web/simplepush, web/webpush, web/registration, web/message, etc. Now it seems the web 'base' is living in both autopush/base.py and autopush/web/base. autopush/base.py was originally only a refactor while endpoint.py lives, ideally it was going to then die so that web/base would contain the base logic for all web handlers.

If the new intent is that all the web handling lives in its own package off the root, we'll need to refactor the refactored webpush/simplepush again for consistent architecture.

jrconlin commented:
I've refactored the refactor to put everything under ./web. I've also gone through the docs and cleaned up pointers to the more critical classes. (Validators aren't included mostly because there aren't many doc strings for them.)

bbangert commented on autopush/web/simplepush.py:
```diff
@@ -2,14 +2,8 @@
-    threaded_validate,
-    SimplePushRequestSchema,
-)
+from autopush.web.base import threaded_validate, Notification, BaseWebHandler
```
3 is enough that we should switch to () multi-line import.

jrconlin replied:
✅

bbangert commented on autopush/web/truss_handler.py:
```diff
@@ -0,0 +1,225 @@
+        if self.valid_input.get('chid'):
+            params["channelID"] = self.valid_input['chid']
+        if "channelID" not in params:
+            params["channelID"] = self.new_id().hex
```
```python
params["channelID"] = self.valid_input.get("chid", self.new_id().hex)
```

jrconlin replied:
✅

bbangert commented on autopush/web/truss_handler.py:
```diff
@@ -0,0 +1,225 @@
+    def new_id(self):
+        """return a UUID4 value
+
+        This is mostly to deal with @patch not always working for testing
```
We have plenty of tests that patch this successfully though? Which test specifically, let's fix that instead of add code like this?

jrconlin replied:
Not a super fan, but yeah, monkeypatching `uuid.uuid4` in the tests also fixes the problem. ✅

bbangert commented on autopush/web/truss_validation.py:
```diff
@@ -0,0 +1,118 @@
+)
+
+
+class TrussSchema(Schema):
```
I don't know what a Truss has to do with Registration. Is there some analogy I'm missing?

jrconlin replied:
Dropping, but Registration* was used by bridges as a support. Thus truss. Granted, we had some file name conflicts that have since gone away, so perfectly fine using registration again.

bbangert commented on autopush/websocket.py:
```diff
@@ -37,11 +37,23 @@
+    has_connected_this_month,
+    hasher,
+    generate_last_connect,
+    dump_uaid)
```
nit: drop the paren down, this looks like pycharm autoformat fail

jrconlin replied:
✅

jrconlin commented on autopush/tests/test_integration.py:
```diff
@@ -1436,6 +1433,7 @@ def _create_context(self, certfile):
 
     @inlineCallbacks
     def test_client_cert_simple(self):
+        self.use_webpush = False
```
✅

bbangert commented on autopush/websocket.py:
```diff
@@ -37,11 +37,23 @@
+from autopush.utils import (
+    parse_user_agent,
+    validate_uaid,
+    WebPushNotification, ms_time)
```
nit: drop the paren down, this looks like pycharm autoformat fail

jrconlin replied:
✅

jrconlin commented on autopush/web/base.py:
```diff
@@ -19,14 +20,90 @@
     401: "Unauthorized",
     404: "Not Found",
     413: "Payload Too Large",
+    418: "I'm a teapot",
```
We generate 418's via our log check. We now check status messages via a handler that looks in this table, so...

jrconlin commented on autopush/main.py:
```diff
@@ -30,10 +27,9 @@
-
+from twisted.internet import reactor, task
+from twisted.logger import Logger
+from twisted.web.server import Site
```
This looks like pycharm may have helped. flake8 also didn't gripe, obviously. 

I'll reshuffle these to our preferred ordering and grouping.

jrconlin commented on autopush/tests/test_health.py:
```diff
@@ -1,5 +1,11 @@
+from autopush.settings import AutopushSettings
+from autopush.web.health import (
+    HealthHandler,
+    MissingTableException,
```
agreed, going to move all exceptions to that module.

jrconlin replied:
✅

jrconlin commented on autopush/tests/test_endpoint.py:
```diff
@@ -34,6 +34,8 @@
     hasher
 )
+from autopush.web.message_handler import MessageHandler
+from autopush.web.truss_handler import RegistrationHandler
```
✅

jrconlin commented on autopush/main.py:
```diff
@@ -2,26 +2,23 @@
 from autopush.logging import PushLogger
 from autopush.settings import AutopushSettings
 from autopush.ssl import AutopushSSLContextFactory
+from autopush.web.health import (HealthHandler, StatusHandler)
```
✅

bbangert commented on autopush/web/registration.py:
```diff
@@ -0,0 +1,311 @@
+
+class RegistrationSchema(Schema):
+    uaid = fields.Str(allow_none=True)
+    chid = fields.Str(allow_none=True)
```
Bummer we can't use UUID objects everywhere, since the UUID object is made below anyways.

jrconlin replied:
We can switch UAID over to UUID, but I think we'll need to keep CHID as a Str since that could be sent by the client.

jrconlin replied:
✅

bbangert commented on autopush/web/registration.py:
```diff
@@ -0,0 +1,311 @@
+            raise InvalidRequest("Invalid router",
+                                 status_code=400,
+                                 errno=108)
+        data['chid'] = data.get('chid')
```
Huh? The chid is definitely there, since pre_load made it, no?

bbangert commented on autopush/web/registration.py:
```diff
@@ -0,0 +1,311 @@
+        # normalize the path vars into parameters
+        router = self.ap_settings.routers[self.valid_input['router_type']]
+
+        if not self.valid_input['uaid']:
```
Couldn't the schema have a default?

bbangert commented on autopush/web/registration.py:
```diff
@@ -0,0 +1,311 @@
+        params = self.valid_input['params']
+        # If the client didn't provide a CHID, make one up.
+        # Note, valid_input may explicitly set "chid" to None
+        params["channelID"] = (self.valid_input.get("chid") or
```
There's multiple things happening in the schema that make it 'for sure' that a chid entry is present, why the get on it here?

bbangert replied:
Why not have a default in the schema?

jrconlin replied:
Because it may not be. The validator (which is called for all REST methods, if I understand correctly) pulls the CHID and verifies if it exists, but it may not since some REST methods note it as optional. POST provides a CHID if it's not present.

bbangert replied:
Ah, makes sense. The 'get(' call shouldn't be needed though, due to the pre_load putting in a chid/uaid key, the valid_input dict should always have those entries.

jrconlin replied:
✅

bbangert commented on autopush/web/registration.py:
```diff
@@ -0,0 +1,311 @@
+            router_type=req['path_kwargs'].get('router_type'),
+            router_token=req['path_kwargs'].get('router_token'),
+            uaid=uaid,
+            chid=chid,
```
Add default entries here?

jrconlin replied:
No. Delete is either the UAID + CHID or just the UAID for a full deregistration.

I suppose I can add special validation handing for the various message types, but I was thinking that it would be clearer to have that in the handlers, where we have to do the logic anyway.

bbangert replied:
Ok, so by specifying the uaid/chid entries here, it is 100% guaranteed that the valid_input will have a key for both of these, no?

bbangert commented on autopush/web/webpush.py:
```diff
@@ -3,14 +3,11 @@
-from autopush.web.validation import (
-    threaded_validate,
+from autopush.web.base import threaded_validate, BaseWebHandler
+from autopush.web.socket_validation import (
```
Why the rename to socket_validation?

jrconlin replied:
:+1:

jrconlin replied:
✅

bbangert commented on autopush/websocket.py:
```diff
@@ -59,23 +74,9 @@
 from twisted.web._newclient import ResponseFailed
 from twisted.web.resource import Resource
+from typing import List  # flake8: noqa
 from zope.interface import implements
```
Looks like another Pycharm PEP8 reformat gone bad.

jrconlin replied:
✅

jrconlin commented on autopush/web/message.py:
```diff
@@ -0,0 +1,71 @@
+from autopush.utils import WebPushNotification
+from cryptography.fernet import InvalidToken
+from marshmallow import Schema, fields, pre_load
+from twisted.internet.threads import deferToThread
```
TIL: File | Settings | Editor > General > Code Folding > [ ] Imports

That should help me spot changes in there.

jrconlin commented on autopush/web/registration.py:
```diff
@@ -0,0 +1,311 @@
+from cryptography.hazmat.primitives import constant_time
+from marshmallow import Schema, fields, pre_load, validates_schema
+from twisted.internet.defer import Deferred
+from twisted.internet.threads import deferToThread
```
✅

jrconlin commented on autopush/web/__init__.py:
```diff
@@ -1 +1,7 @@
+bridge registration endpoints, health, and error testing endpoints. Many of
+these endpoints run through marshmallow validation.
+
+:"""
```
✅

bbangert commented on autopush/router/apnsrouter.py:
```diff
@@ -7,7 +7,8 @@
 from twisted.logger import Logger
 from twisted.internet.threads import deferToThread
-from autopush.router.interface import RouterException, RouterResponse
+from autopush.router.interface import RouterResponse
```
Should be a gap here per PEP8.

bbangert commented on autopush/tests/test_health.py:
```diff
@@ -1,5 +1,11 @@
+from autopush.web.health import (
+    HealthHandler,
+    StatusHandler,
+)
```
PEP8

bbangert commented on autopush/tests/test_integration.py:
```diff
@@ -13,27 +13,28 @@
 )
 from autopush.settings import AutopushSettings
 from autopush.utils import base64url_encode
+from autopush.main import endpoint_paths
```
Ideally these are mostly alphabetical in order.

bbangert commented on autopush/tests/test_log_check.py:
```diff
@@ -1,18 +1,17 @@
+from moto import mock_dynamodb2
+from nose.tools import eq_
 from twisted.internet.defer import Deferred
+from twisted.trial import unittest
```
PEP8

jrconlin replied:
I'm not sure how this violates PEP8. It's a group of non-local imports, ordered with a break before the local modules. It's a single entry, so no grouping is required.

bbangert commented on autopush/web/log_check.py:
```diff
@@ -1,23 +1,29 @@
+    Schema,
+    fields,
+    pre_load
+)
```
PEP8

bbangert commented on autopush/web/push_validation.py:
```diff
@@ -2,8 +2,20 @@
+from autopush.utils import (
+    base64url_encode,
+    extract_jwt,
+    WebPushNotification)
```
PEP8
